google/cloud/forseti/common/opencensus/tracing.py

```diff
@@ -0,0 +1,347 @@
+    integrated_libraries = config_integration.trace_integrations(
+        integrations,
+        tracer)
+    LOGGER.info(f'Tracing libraries: {integrated_libraries}')
+    return integrated_libraries
+
+
+def create_exporter(transport=None):
+    """Create an exporter for traces.
+
+    The default exporter is the StackdriverExporter. If it fails to initialize,
+    the FileExporter will be used instead.
+
+    Args:
+        transport (opencensus.trace.common.transports.base.Transport): the
+            OpenCensus transport used by the exporter to emit data.
+
+    Returns:
+        StackdriverExporter: A Stackdriver exporter.
+        FileExporter: A file exporter. Default path: 'opencensus-traces.json'.
+    """
+    transport = transport or AsyncTransport
+    try:
+        exporter = stackdriver_exporter.StackdriverExporter(transport=transport)
+        LOGGER.info(
+            'StackdriverExporter set up successfully for project %s.',
+            exporter.project_id)
+        return exporter
+    except Exception:  # pylint: disable=broad-except
+        LOGGER.exception(
+            'StackdriverExporter set up failed. Using FileExporter.')
+        return file_exporter.FileExporter(transport=transport)
+
+
+def traced(methods=None, attr=None):
+    """Class decorator to enable automatic tracing on designated class methods.
+
+    Args:
+        methods (list, optional): If set, the decorator will trace those class
+            methods. Otherwise, trace all class methods.
+        attr (str, optional): If the tracer was passed explicitly to the class
+            in an attribute, get it from there.
+
+    Returns:
+        object: Decorated class.
+    """
+    def wrapper(cls):
+        """Decorate selected class methods.
+
+        Args:
+            cls (object): Class to decorate.
+
+        Returns:
+            object: Decorated class.
+        """
+        # Get names of methods to be traced.
+        cls_methods = inspect.getmembers(cls, inspect.ismethod)
+        if methods is None:
+            # trace all class methods
+            to_trace = cls_methods
+        else:
+            # trace specified methods
+            to_trace = [m for m in cls_methods if m[0] in methods]
+
+        # Decorate each of the methods to be traced.
+        # Adds `self.tracer` in class to give access to the tracer from within.
+        if OPENCENSUS_ENABLED:
+            for name, func in to_trace:
+                LOGGER.info(f'Tracing - Adding decorator to {name}')
+                if name == '__init__':
+                    # __init__ decorator to add tracer as instance attribute
+                    decorator = trace_init(attr=attr)
+                else:
+                    decorator = trace()
+                setattr(cls, name, decorator(func))
+        return cls
+    return wrapper
+
+
+# pylint: disable=redundant-returns-doc
+def trace_init(attr=None):
+    """Method decorator for a class's __init__ method. Set `self.tracer` (either
```

Comment:
Summary line of the docstring should be one line only.  If this is longer, how about:
```
"""Method decorator for a class's __init__ method. 

Set `self.tracer` (either from instance kwargs, attribute, or execution context).
```

Reply:
Updated.